Rework racket based runtime/compilation

dolio opened the pull request:
This PR includes the changes that get us about as far as we can easily go for our racket evaluation/compilation at the moment. It includes the following aspects.

- I renamed the `runner` file/executable to `ucr`, so that it's more similar to `ucm`. It's not something that people should be calling directly, but `runner` is obviously not a very good name.
- I changed the 'compilation' part of `ucr` to just generate a file. The reason for this is that somehow a math library we are using for the primops causes `create-embedding-executable` to fail. The easiest solution to this problem is to divide the compilation process into two steps: first generate the file using all the unison infrastructure, then compile that file. At first I was thinking about copying the math operations into the primops, but I realized that dividing the programs up is a better solution.
- At the moment, compiling the generated file is just done via `raco exe` using a racket installation. The reason for _this_ is that `create-embedding-executable` does not seem to work in a standalone executable. The API needs to find a copy of the racket executable to modify, and the procedure for that doesn't work in a standalone binary (confirmed with racket devs). Peeking at the code, it doesn't seem like this absolutely _couldn't_ work, but making it work would require someone digging into the racket internals to make it happen. So for the time being, compiling requires a racket install.
- I've included a short transcript that will download `base`, `@unison/internal`, and generate the racket modules alongside the checked-in modules. This is a step necessary in building `ucr` or getting things set up for compilation, so we'll probably use it when we get this stuff building in CI.

Even for compilation, where we still depend on a racket install, we no longer actually need to make people download `@unison/internal` themselves. We can run the library generation as part of a build, and just distribute the generated libraries to people. The runtime should be completely self-contained, though. Once `ucr` is built, `ucm` just needs to shell out to it to execute `run.native`. A racket install is only necessary to _build_ `ucr`.

I think one tweak I have left is a better error message for `compile.native` if `raco` doesn't exist. Beyond that, I just need a little advice on how to get this hooked into CI to get everything distributed properly. I might need to make the native runtime calls a bit more configurable with respect to the location of `ucr` and such, too, but that shouldn't be a big deal.

dolio committed:
Further work on compilation to standalone executables

Switch compile.native to using the native Runtime

Tweak the call for compilation to use `racket runner.rkt`, since it
seems like _compiling_ runner.rkt precludes the use of
create-embedding-executable, at least until something further is figured
out.

runner and other racket infrastructure additions for producing the right
input for racket executable creation

dolio committed:
Revise how native compilation works

There are, for the moment, multiple problems with producing executables.
One is that the API doesn't work when certain math libraries are
imported. The other is that the API doesn't work in a generated
executable.

So, I've turned the 'compile mode' of the standalone runtime program
just generate a module to be compiled. Then we can use an installed
`raco` to build that module into a standalone program. I don't think
there's any advantage currently to not using raco, since the best we can
do is write our own separate exe builder that still requires racket to
be installed.

In any case, the exe builder must apparently be a separate program so
that it doesn't need to import the offending math library.

dolio committed:
Rename runner, add docs, delete dead code

dolio committed:
Add a transcript to create generated racket libs

aryairani commented:
> * I renamed the `runner` file/executable to `ucr`, so that it's more similar to `ucm`. It's not something that people should be calling directly, but `runner` is obviously not a very good name.

I'm thinking we could use a more descriptive name and stick it in `libexec` if nobody needs to call it.

> * I changed the 'compilation' part of `ucr` to just generate a file.

What did it do before?

> * I've included a short transcript that will download `base`, `@unison/internal`, and generate the racket modules alongside the checked-in modules. This is a step necessary in building `ucr` or getting things set up for compilation, so we'll probably use it when we get this stuff building in CI.
> 
> Even for compilation, where we still depend on a racket install, we no longer actually need to make people download `@unison/internal` themselves. We can run the library generation as part of a build, and just distribute the generated libraries to people. The runtime should be completely self-contained, though. Once `ucr` is built, `ucm` just needs to shell out to it to execute `run.native`. A racket install is only necessary to _build_ `ucr`.

Can we just do all of this in CI?  Build `ucr` there, and nobody needs to install racket outside of CI?

> I think one tweak I have left is a better error message for `compile.native` if `raco` doesn't exist. Beyond that, I just need a little advice on how to get this hooked into CI to get everything distributed properly. 

We can pair on the CI part if you want.

aryairani commented on unison-cli/src/Unison/Codebase/Editor/HandleInput.hs:
```diff
@@ -2071,21 +2071,21 @@ runScheme file args = do
-        catch
-          (True <$ callProcess "racket" opts)
-          (\(_ :: IOException) -> pure False)
+-- buildScheme :: String -> String -> Cli ()
+-- buildScheme main file = do
+--   ensureSchemeExists
+--   statDir <- getSchemeStaticLibDir
+--   genDir <- getSchemeGenLibDir
+--   buildRacket genDir statDir main file
+
+-- buildRacket :: String -> String -> String -> String -> Cli ()
+-- buildRacket genDir statDir main file =
+--   let args = ["-l", "raco", "--", "exe", "-o", main, file]
+--       opts = racketOpts genDir statDir args
+--    in void . liftIO $
+--         catch
+--           (True <$ callProcess "racket" opts)
+--           (\(_ :: IOException) -> pure False)
```
Keep these in for now for reference? Or accidental inclusion?

dolio replied:
Just haven't deleted them yet.

aryairani commented on unison-cli/src/Unison/Codebase/Editor/HandleInput.hs:
```diff
@@ -2111,9 +2111,9 @@ doRunAsScheme main0 args = case HQ.fromString main0 of
-doCompileScheme :: String -> HQ.HashQualified Name -> Cli ()
-doCompileScheme out main =
-  generateSchemeFile True out main >>= buildScheme out
+-- doCompileScheme :: String -> HQ.HashQualified Name -> Cli ()
+-- doCompileScheme out main =
+--   generateSchemeFile True out main >>= buildScheme out
```
Same

dolio commented:
> I'm thinking we could use a more descriptive name and stick it in `libexec` if nobody needs to call it.

Any suggestion on the name? I guess it could be something more like `unison-runtime` or something.

> What did it do before?

It generated a file then threw errors. But the command line gave you the expectation that it would produce an executable.

> Can we just do all of this in CI? Build `ucr` there, and nobody needs to install racket outside of CI?

You can do the equivalent of `run`. Or you can generate a racket source file, but that's useless without racket. If you want to build a standalone executable from some unison code, you need racket installed, because the API for generating executables doesn't work when you build it into a standalone executable.

dolio committed:
Give a nicer looking error for ucr/raco problems

dolio committed:
Eliminate commented dead code

dolio committed:
Add an improved error message for unison native runtime failure

dolio committed:
Add information on further building to gen-racket-libs.md

ChrisPenner commented on parser-typechecker/src/Unison/Runtime/Interface.hs:
```diff
@@ -696,23 +704,43 @@ nativeEvalInContext _ ctx codes base = do
+      ucrError (_ :: IOException) =
+        die
+          "I had trouble calling the unison runtime exectuable.\n\n\
+          \Please check that the `ucr` executable is properly\
```
Would be great to link to a guide in these errors, as an end user I wouldn't know where to start on ensuring "ucr" is installed and google probably won't help.

dolio replied:
Yeah, I'd like to make them better. I'm not exactly sure what the install process will be right now, though. Hopefully I can flesh these out after I work out some details with Arya.

aryairani replied:
I'm anticipating that either our installers will set it up correctly for them, or they're in big trouble and need to come ask for help, which will help us flesh out better error messages :-\

ChrisPenner commented:
Looks great! I'd second the idea of changing `ucr` to something like `unison-runtime`

dolio committed:
Actually make run.native use the standalone runner

All the infrastructure was there already, just needed to enable it.

This should break the jit tests in CI until we can get the runner
building as a prior step, and coordinate the location the native runtime
looks for the build files.

dolio committed:
Merge remote-tracking branch 'origin/trunk' into topic/native-compiler

dolio committed:
Rename ucr to unison-runtime

dolio committed:
Add info file for unison racket package

dolio committed:
Native runtime location argument, and an improved error msg

aryairani commented:
When would `racoErrMsg` show up currently? Current plan is to never have to call it from within `ucm`, right?  (It doesn't need to be deleted yet, but maybe once CI is working?)

Commit:
refactor ci.yaml with jit setup steps (attempt 1)

recommend moving support scheme files to another repo
if they aren't deeply related to ucm; but maybe they are.

Commit:
yaml

Commit:
unyaml

Commit:
yaml home

Commit:
yaml unhome

Commit:
yaml move env

Commit:
yaml matrix

Commit:
yaml can't define vars in terms of vars

but bash tricks can maybe?

Commit:
yaml quotes sometimes

Commit:
yaml renamed var

Commit:
yaml need env namespace

Commit:
yaml broken variables hack

Commit:
yaml hack documentation

Commit:
yaml job steps

Commit:
yaml tweak identifier?

Commit:
yaml that didn't work

Commit:
yaml bump plugins to node 20

dolio committed:
Modify transcript runner to allow runtime path

Fixes problems with it building, since it hadn't been updated to call
other modified functions in the right way.

dolio committed:
Merge remote-tracking branch 'origin/topic/native-compiler' into topic/native-compiler

dolio commented:
`racoErrMsg` shows up when you run `compile.native` and racket isn't installed. You need racket to compile standalone executables right now.

dolio committed:
Fix ucm tests w/r/t/ previous additions

Commit:
yaml can't evaluate shell expression as value

Commit:
Merge branch 'topic/native-compiler' of github.com:unisonweb/unison into topic/native-compiler

Commit:
yaml step id

Commit:
yaml update deprecated output convention

Commit:
yaml debug windows artifact failure

Commit:
yaml temporarily disable tests

Commit:
yaml failing to uploading files is still success

Commit:
yaml swearing

Commit:
yaml whoops

Commit:
yaml swearing+1

Commit:
yaml maybe move bins somewhere nicer

Commit:
yaml getting desperate

Commit:
yaml loving magical filenames

Commit:
yaml unmagic filenames. are duplicate ids ok?

Commit:
yaml no good, does globbing work for upload-artifact?

Commit:
yaml test linux only, for a change

Commit:
yaml globs don't work

Commit:
yaml copilot whoops

Commit:
skip tmate for now

Commit:
yaml blowing up because we skipped ormolu?

Commit:
yaml slightly better?

Commit:
Merge branch 'trunk' into topic/native-compiler

# Conflicts:
#	unison-cli/unison/ArgParse.hs
#	unison-cli/unison/Main.hs

Commit:
yaml can i get an `if` that makes any sense

Commit:
comma

Commit:
yaml switch plugin

Commit:
yaml x2

Commit:
yaml get scheme-libs/

Commit:
yaml missed one variable name swap

Commit:
yaml where are my files?

Commit:
yaml make temp scheme dir before

Commit:
yaml downloaded unison artifact filename is just called unison

Commit:
yaml maybe transcript was being overwritten by checkout

Commit:
yaml create transcript with separate path and file

Commit:
yaml try tmate again

Commit:
yaml i hope quotes don't matter

Commit:
yaml try finnp/create-file-action@v2 instead?

Commit:
try DamianReeves/write-file-action@v1.3 instead

Commit:
yaml don't need to run transcript here

Commit:
yaml untmate

Commit:
commit racket deps list

Commit:
yaml no apt on windows

Commit:
yaml tmate

Commit:
yaml forget broken files-exist action

Commit:
yaml skip clone if src not needed

Commit:
yaml always save jit source for next step (for now)

Commit:
yaml more yaml

Commit:
yaml all platforms without tmate

Commit:
yaml set up base codebase

Commit:
don't fetch internal

Commit:
yaml typo

Commit:
yaml stop before jit tests

Commit:
yaml mac doesn't have apt either

Commit:
yaml switch windows path separator

Commit:
yaml file existence plugins don't work

Commit:
yaml the lack of proper flow control is astounding

Commit:
yaml always checkout repo prior to running transcript from it

Commit:
yaml whoops

Commit:
yaml no clue whats going on again

Commit:
yaml remove caching condition

Commit:
yaml works better when uncommented

Commit:
yaml does overriding the default with the default make it work better?

Commit:
yaml lame that that worked

Commit:
yaml try not to blow up on the second run

Commit:
yaml argh

Commit:
yaml disable pause again

Commit:
yaml why did windows fail to build jit binary?

Commit:
yaml because of course you have to specify always

Commit:
yaml jump through some hoops re `.exe` on windows

Commit:
yaml small refactor, hope I don't regret it

Commit:
yaml it never ends

Commit:
jit tests reply on some types not in base

at least ThreadKilledFailure, from https://github.com/unisonweb/unison/pull/3835

Commit:
all the mergeio types we rely on should be in base I suppose?

dolio committed:
Make --runtime-path expect full executable path

Commit:
yaml change semantics of --runtime-path

Commit:
Merge branch 'topic/native-compiler' of github.com:unisonweb/unison into topic/native-compiler

Commit:
yaml definitely almost done

Commit:
yaml fix jit source cache key

Commit:
raco distribute builds a different directory structure in windows than in mac/linux

Commit:
yaml accidentally a )

Commit:
yaml accidentally the other )

Commit:
yaml actually wanted jit version, not jit source hash which we don't know

Commit:
yaml same here

Commit:
simplify but hopefully not break caching

Commit:
add ioexception message to ucr and raco failures.

Commit:
yaml add windows-2022 os

Commit:
cache tips

Commit:
yaml maybe less broken caching

Commit:
beef up the error messages around calling the native runtime

Commit:
some reference info

Commit:
do re-save ~/.stack if package dependencies have changed

Commit:
do re-save `.stack-work` after build if Haskell files have changed

Commit:
more debug info for calling raco

Commit:
yaml nothing useful from windows-2022 after all

Commit:
yaml set 15 minute tmate timeout

Commit:
try to restore jit binaries before checking for them

Commit:
yaml reenable all the ucm tests

figure out how to cache them later

Commit:
yaml whoops

Commit:
another yaml reference

aryairani commented:
Heads up that I'm seeing some intermittent failures in the interpreter tests on macOS.
https://github.com/unisonweb/unison/actions/runs/7937787444/job/21675781417?pr=4675#step:16:173

Any chance that's related to this PR? Otherwise I can hope for green and open a separate ticket.

Commit:
yaml whoops

how would it be to be notified of a syntax error before uploading the thing

aryairani commented:
Note to self: Currently the interpreter tests are being run in the "build JIT binary" stage, which is a little wrong.
But also maybe that's fine / easier because they both rely on the result of `base.md` and it might be a pain to factor that out

Commit:
yaml simplify and hopefully not break our diffs

Commit:
yaml move ucm local bin path out of repo?

Commit:
yaml hopefully skip hopefully unnecessary work

Commit:
yaml rerun if transcripts have changed

Commit:
yaml fixup windows slashes again

Commit:
yaml case sensitive env var names

Commit:
yaml blerk

Commit:
yaml skip failing interpreter tests in windows for now

Commit:
caching tips

Commit:
skip jit integration test on windows

Commit:
yaml: `timeout-minutes:` is a step setting, not an actions setting.

Commit:
yaml: skip base pull on cache hit

Commit:
yaml move restore jit binaries step relative to environment tweak

Commit:
yaml switch environment tweak relative to restore jit binaries step

Commit:
update github-actions-help

Commit:
yaml: error if nothing to upload

Commit:
yaml: work around https://github.com/actions/cache/issues/1334

aryairani reviewed (changes requested):
There's no way to remove an approval (just to make sure it's not merged prematurely); so I'm marking it as requesting changes.

I'm still working on updated drafts of the release scripts, and Dan is changing the behavior of `Nat`.

aryairani reviewed (changes requested):
There's no way to remove an approval (just to make sure it's not merged prematurely); so I'm marking it as requesting changes.

I'm still working on updated drafts of the release scripts, and Dan is changing the behavior of `Nat`.

Commit:
yaml: will `--embed-dlls` allow jit integration test to succeed on windows?

Commit:
yaml timeout was too short for debugging

Commit:
update github-actions-help

Commit:
commend

dolio committed:
Racket chunked seq fixes for base tests

dolio committed:
Add additional universal compare cases in racket

dolio committed:
Define some helper functions for modular arithmetic

These attempt to put slow operations behind a branch, so that arithmetic
on small enough numbers can remain efficient.

dolio committed:
Implement modular arithmetic for Int/Nat

dolio committed:
Scope.bytearrayOf and missing imports from prior commit

dolio committed:
Racket pattern fixes

dolio committed:
Add a version flag to unison-runtime and rework slightly

dolio committed:
Fix char class stuff (more) properly

dolio committed:
Fix racket float to text formatting a bit to match Haskell

dolio committed:
Tweak fromUtf8 error message; add IOFailure type in scheme

Requires bumping the jit implementation version info

dolio committed:
Merge remote-tracking branch 'origin/trunk' into topic/native-compiler

dolio commented:
Okay, I think the tests in `@aryairani/jit-tests` should pass with this branch.

I also added `compiler.generateSchemeBoot` to `@unison/internal` in the new release `0.0.11` that this depends on. I think you were planning on doing something with that in the transcripts?

dolio committed:
Bump gen-racket-libs version for @unison/internal

aryairani committed:
Update ci.yaml with jit version 0.0.11

Commit:
disable windows jit integration tests until #4705

Commit:
yaml: temporarily print what's in the cached system stack dir

Commit:
yaml: add test caching and change a bunch of caching stuff

and hope for the best

Commit:
yaml: forgot a plugin name

Commit:
yaml: temporarily print what's in the stack cache, better

Commit:
yaml: fix test results caching?

tests results are cached with the ucm binaries

Commit:
yaml: bikeshed variable names and hope nothing broke

Commit:
yaml: these were supposed to be part of a previous commit

Commit:
yaml: call generateSchemeBoot directly from `@unison/internal`

Commit:
yaml: cache key vars for invalidation

aryairani reviewed (approved):
I'm gonna doublecheck CI again in the morning, but I think this is ready to merge even though we may do some additional tweaks later.

aryairani commented:
> Okay, I think the tests in `@aryairani/jit-tests` should pass with this branch.
> 
> I also added `compiler.generateSchemeBoot` to `@unison/internal` in the new release `0.0.11` that this depends on. I think you were planning on doing something with that in the transcripts?

Thanks for the reminder, got it in 20a0e73.

Commit:
CI docs

aryairani committed:
Update ci.md

aryairani committed:
Update ci.md

aryairani committed:
Update ci.md

Commit:
yaml: clarify that you shouldn't have to invalidate build caches

Commit:
Merge branch 'topic/native-compiler' of github.com:unisonweb/unison into topic/native-compiler

Commit:
remove `compile.native.{fetch,genlibs}` commands

existing ucm commands can be used instead, i.e. `gen-racket-libs.md`

aryairani commented:
~The default runtime location is set in `fixNativeRuntimePath`, currently as `$XDG_DATA/unisonlanguage/libexec/unison-runtime`, and can be overridden by `nativeRuntimePathFlag`.~

- [x] It should probably default to a system-wide location though, like `$0/unison-runtime`?

The default runtime location is set in `fixNativeRuntimePath`, as `$0/runtime/unison-runtime[.exe]`, and can be overridden by `nativeRuntimePathFlag`.

aryairani commented:
> A racket install is only necessary to build `ucr`.

The racket install is still needed to use `compile.native`, right?

Commit:
yaml: update stack caching

Commit:
do we need `base.output.md`?

Commit:
change default ucr location to `$0/runtime/unison-runtime[.exe]`

Commit:
yaml: maybe fix stack caching condition again?

Commit:
yaml: predicate requires British spelling
